returnn/tf/util/data.py

```diff
@@ -1190,123 +1243,57 @@ def copy_compatible_to(self, data, unbroadcast=False, except_feature=False,
+    mapped_axes = data.find_matching_dim_map(v, list(range(v.batch_ndim)), is_equal_opts)  # maps v -> data
+    assert len(mapped_axes) == v.batch_ndim
+
+    for target_axis in range(data.batch_ndim):
+      new_v_axis = min(target_axis, v.batch_ndim)
+      if target_axis not in mapped_axes.values():
+        # Dim in data, but not in v
+        unbroadcast_axis = unbroadcast and not (except_feature and data.feature_dim_axis == target_axis)
+        v = v.copy_add_dim_by_tag(data.get_dim_tag(target_axis), axis=new_v_axis, unbroadcast=unbroadcast_axis)
+        # Keep mapped_axes consistent
+        mapped_axes = {v_ax + (1 if v_ax >= new_v_axis else 0): trg_ax for v_ax, trg_ax in mapped_axes.items()}
+        mapped_axes[new_v_axis] = target_axis
+      else:
+        # Dim exists in both data and in v. Maybe order is wrong.
+        matching_v_axes = [v_ax for v_ax, trg_ax in mapped_axes.items() if trg_ax == target_axis]
+        assert len(matching_v_axes) == 1
+        matching_v_axis = matching_v_axes[0]
+        if target_axis != matching_v_axis:
+          # Order was wrong
+          v = v.copy_swap_axes(matching_v_axis, new_v_axis)
+          # Keep mapped_axes consistent
+          mapped_axes[matching_v_axis], mapped_axes[new_v_axis] = mapped_axes[new_v_axis], mapped_axes[matching_v_axis]
+
     assert v.batch_ndim == data.batch_ndim
-    # Now maybe move batch/feature axis.
-    # We might do multiple iterations here, depending on which axis comes first.
-    # This is a bit ugly, but the code is simpler.
-    num_iterations = 0
-    while True:
-      num_iterations += 1
-      assert num_iterations <= 4
-      if v.batch_dim_axis != data.batch_dim_axis:
-        assert data.batch_dim_axis is not None and v.batch_dim_axis is not None
-        v = v.copy_with_batch_dim_axis(data.batch_dim_axis)
-        assert v.batch_dim_axis == data.batch_dim_axis
-        continue
-      if v.feature_dim_axis != data.feature_dim_axis:
-        assert data.feature_dim_axis is not None and v.feature_dim_axis is not None
-        v = v.copy_with_feature_dim_axis(data.feature_dim_axis)
-        assert v.feature_dim_axis == data.feature_dim_axis
-        if data.feature_dim_axis_or_unspecified is NotSpecified:
-          v.feature_dim_axis = NotSpecified
-          assert v.feature_dim_axis == data.feature_dim_axis
-        continue
-      # Now we have both equal.
-      break
-    if data.feature_dim_axis_or_unspecified is NotSpecified and v.feature_dim_axis_or_unspecified is not NotSpecified:
-      if v._default_feature_dim_axis() == v.feature_dim_axis:
-        v.feature_dim_axis = NotSpecified
+    assert all(mapped_axes[ax] == ax for ax in range(v.batch_ndim))
+
+    # Ensure time_dim_axis and feature_dim_axis is same as in data
+    assert v.batch_dim_axis == data.batch_dim_axis  # there is only at most one batch_dim_axis
+    v.time_dim_axis = data.time_dim_axis
+    if data.feature_dim_axis_or_unspecified is NotSpecified:
```

Comment:
Doesn't `feature_dim_axis_or_unspecified` handle both cases? So just `v.feature_dim_axis = data.feature_dim_axis_or_unspecified`?

Reply:
Yes right, that would be simpler. @Zettelkasten please also change that.